### PR TITLE
Quarkus source S2I strategy

### PR DESCRIPTION
**Pull request by mjurc**
* Adding Quarkus source S2I strategy. This strategy utilises OpenShift
  objects generated by the following oc commands:

   ```
   oc import-image --confirm ubi8/openjdk-11 --from=registry.access.redhat.com/ubi8/openjdk-11
   oc new-app ubi8/openjdk-11 <git_path> --context-dir=<context_dir> --name=<project_name>
   ```

  There are additional enhancements, like adding environment variables,
  which are not available from `oc new-app` command.

### CODE REVIEW

**Comment by mjurc** [comment]
@Sgitario I'm reopening this over https://github.com/quarkus-qe/quarkus-test-framework/pull/32 as the fork relationships have been lost in my release tryouts on my fork and the PR was not getting updated.

File: examples/external-applications/src/test/java/io/quarkus/qe/OpenShiftS2iQuickstartIT.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+import org.apache.http.HttpStatus;
+import org.junit.jupiter.api.Test;
+
+import io.quarkus.test.bootstrap.RestService;
+import io.quarkus.test.scenarios.OpenShiftDeploymentStrategy;
+import io.quarkus.test.scenarios.OpenShiftScenario;
+import io.quarkus.test.services.QuarkusApplication;
+
+@OpenShiftScenario(deployment = OpenShiftDeploymentStrategy.QuarkusS2IBuild)
```
Comment (mjurc) [inline examples/external-applications/src/test/java/io/quarkus/qe/OpenShiftS2iQuickstartIT.java:11]: @Sgitario Actually I am not sure how to achieve this. What comes to mind is moving the logic from `QuarkusSourceS2iBuildApplicationManagedResource` to `BuildOpenShiftQuarkusApplicationManagedResource` and forking it based on presence of `QuarkusAnnotation` properties, but if there's a better way, I'm all ears!

Comment (Sgitario) [reply examples/external-applications/src/test/java/io/quarkus/qe/OpenShiftS2iQuickstartIT.java:11]: I like your idea! The question is... do you want to do it as part of this PR or in a second iteration of this feature?

File: quarkus-test-openshift/src/main/resources/quarkus-s2i-source-build-template.yml [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        type: Git
+        configMaps:
+          - configMap:
+              name: settings-mvn
+            destinationDir: "/configuration"
+      strategy:
+        type: Source
+        sourceStrategy:
+          env:
+            - name: ARTIFACT_COPY_ARGS
```
[the remark this answers is not in the document]
Comment (Sgitario) [reply quarkus-test-openshift/src/main/resources/quarkus-s2i-source-build-template.yml:33]: Keeping this property in here would make it fail with 2.x (as no runner jar will be found).

**Review by Sgitario — CHANGES_REQUESTED** [review]

File: quarkus-test-core/src/main/java/io/quarkus/test/utils/FileUtils.java [hunk 3]
```diff
@@ -100,4 +102,13 @@ public static void deleteDirectory(Path folder) {
 
         return Optional.empty();
     }
+
+    public static void copyResourceToTargetPath(String resourceFileName, Path targetPath) {
```
[the remark this answers is not in the document]
Comment (mjurc) [reply quarkus-test-core/src/main/java/io/quarkus/test/utils/FileUtils.java]: Updated

File: quarkus-test-openshift/src/main/java/io/quarkus/test/bootstrap/inject/OpenShiftClient.java [hunk 4]
```diff
@@ -460,6 +470,77 @@ private String enrichTemplate(Service service, String template) {
...
+            Serialization.yamlMapper().writeValue(os, list);
+            template = new String(os.toByteArray());
+        } catch (IOException e) {
+            fail("Failed adding properties into OpenShift template. Caused by " + e.getMessage());
+        }
+
+        return template;
+    }
+
+    public String addEnvVarsToDeploymentConfig(String deploymentConfigName, String[] envVars, String template) {
```
[the remark this answers is not in the document]
Comment (mjurc) [reply quarkus-test-openshift/src/main/java/io/quarkus/test/bootstrap/inject/OpenShiftClient.java]: Updated

File: quarkus-test-openshift/src/main/java/io/quarkus/test/bootstrap/inject/OpenShiftClient.java [hunk 5]
```diff
@@ -505,7 +586,7 @@ private boolean isResource(String key) {
     }
 
     private boolean hasImageStreamTags(ImageStream is) {
-        return !masterClient.imageStreams().withName(is.getMetadata().getName()).get().getSpec().getTags().isEmpty();
+        return (!masterClient.imageStreams().withName(is.getMetadata().getName()).get().getStatus().getTags().isEmpty());
```
[the remark this answers is not in the document]
Comment (mjurc) [reply quarkus-test-openshift/src/main/java/io/quarkus/test/bootstrap/inject/OpenShiftClient.java]: Fixed

File: quarkus-test-openshift/src/main/java/io/quarkus/test/services/quarkus/QuarkusSourceS2iBuildApplicationManagedResource.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+                .replaceAll(quote("${GIT_URI}"), model.getGitRepositoryUri())
+                .replaceAll(quote("${QUARKUS_VERSION}"), model.getQuarkusBuildVersion());
+        if (StringUtils.isNotEmpty(model.getContextDir())) {
+            result = client.addContextDirToBuildConfig(model.getContext().getOwner().getName(), model.getContextDir(),
+                    result);
+        }
+        if (StringUtils.isNotEmpty(model.getGitRef())) {
+            result = client.addGitRefToBuildConfig(model.getContext().getOwner().getName(), model.getGitRef(), result);
+        }
+        return result;
```
[the remark this answers is not in the document]
Comment (mjurc) [reply quarkus-test-openshift/src/main/java/io/quarkus/test/services/quarkus/QuarkusSourceS2iBuildApplicationManagedResource.java:111]: Indeed, it seems it does matter - this string is processed by one of enriching methods downstream, resulting in removal of fields without values

File: quarkus-test-openshift/src/main/resources/quarkus-s2i-source-build-template.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (mjurc) [reply quarkus-test-openshift/src/main/resources/quarkus-s2i-source-build-template.yml:35]: Right - you think I should remove it for now, then? I haven't tried the Quickstarts with Quarkus 2 alphas yet.

File: examples/external-applications/src/test/java/io/quarkus/qe/OpenShiftS2iQuickstartIT.java (hunk 1 above)
Comment (mjurc) [reply examples/external-applications/src/test/java/io/quarkus/qe/OpenShiftS2iQuickstartIT.java:11]: I'd prefer next iteration!

Issue: https://github.com/quarkus-qe/quarkus-test-framework/issues/59

File: quarkus-test-openshift/src/main/resources/quarkus-s2i-source-build-template.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (rsvoboda) [reply quarkus-test-openshift/src/main/resources/quarkus-s2i-source-build-template.yml:35]: There is CR1 already :)

File: quarkus-test-core/src/main/java/io/quarkus/test/services/quarkus/ProdQuarkusApplicationManagedResourceBuilder.java [hunk 7]
```diff
@@ -50,6 +50,10 @@ public void init(Annotation annotation) {
         QuarkusApplication metadata = (QuarkusApplication) annotation;
         setSslEnabled(metadata.ssl());
         initAppClasses(metadata.classes());
+        setGitRepositoryUri(metadata.gitRepositoryUri());
+        setGitRef(metadata.gitRef());
+        setContextDir(metadata.contextDir());
+        setQuarkusBuildVersion(metadata.quarkusBuildVersion());
```
Comment (Sgitario) [inline quarkus-test-core/src/main/java/io/quarkus/test/services/quarkus/ProdQuarkusApplicationManagedResourceBuilder.java:56]: Quarkus version should be taken from the Maven configuration. I've created an issue to fix this in a separated PR: https://github.com/quarkus-qe/quarkus-test-framework/issues/65

File: quarkus-test-openshift/src/main/resources/settings-mvn.yml [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+                        <url>https://maven.repository.redhat.com/ga/</url>
+                    </pluginRepository>
+                </pluginRepositories>
+            </profile>
+            <profile>
+                <id>customRemoteRepository</id>
+                <repositories>
+                    <repository>
+                        <id>ts.s2i.maven.remote.repository</id>
+                        <url>${ts.s2i.maven.remote.repository}</url>
```
Comment (Sgitario) [inline quarkus-test-openshift/src/main/resources/settings-mvn.yml:34]: This property must be documented in the README.md. Created https://github.com/quarkus-qe/quarkus-test-framework/issues/66 for doing this.

File: quarkus-test-openshift/src/main/java/io/quarkus/test/bootstrap/inject/OpenShiftClient.java [hunk 9]
```diff
@@ -460,6 +469,50 @@ private String enrichTemplate(Service service, String template) {
...
+            Serialization.yamlMapper().writeValue(os, list);
+            template = new String(os.toByteArray());
+        } catch (IOException e) {
+            fail("Failed adding properties into OpenShift template. Caused by " + e.getMessage());
+        }
+
+        return template;
+    }
+
+    public String addContextDirToBuildConfig(String buildConfigName, String contextDir, String template) {
```
Comment (Sgitario) [inline quarkus-test-openshift/src/main/java/io/quarkus/test/bootstrap/inject/OpenShiftClient.java:494]: Methods addGitRefToBuildConfig and addContextDirToBuildConfig should not be exposed in the OpenShiftClient API. Created https://github.com/quarkus-qe/quarkus-test-framework/issues/67

File: quarkus-test-openshift/src/main/resources/quarkus-s2i-source-build-template.yml [hunk 10]
```diff
@@ -0,0 +1,97 @@
+apiVersion: v1
```
Comment (Sgitario) [inline quarkus-test-openshift/src/main/resources/quarkus-s2i-source-build-template.yml:1]: We should allow providing a custom S2i for other usages. Created https://github.com/quarkus-qe/quarkus-test-framework/issues/68 for doing this.

File: quarkus-test-openshift/src/main/resources/quarkus-s2i-source-build-template.yml [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+        type: Git
+        contextDir: ${CONTEXT_DIR}
+        configMaps:
+          - configMap:
+              name: settings-mvn
+            destinationDir: "/configuration"
+      strategy:
+        type: Source
+        sourceStrategy:
+          env:
```
Comment (Sgitario) [inline quarkus-test-openshift/src/main/resources/quarkus-s2i-source-build-template.yml:34]: I've created another issue to allow configuring these properties https://github.com/quarkus-qe/quarkus-test-framework/issues/69

**Review by Sgitario — APPROVED** [review]
Let's merge this PR. The most important remaining bit is about the Quarkus version which at the moment is not behaving as others. But let's fix that in a separate PR.

**Pull request merged**